### PR TITLE
Feature/background and bounding box color customization

### LINKED ISSUE #31: Add a color picker to select color for Background and Bounding Box
## Use Case

The viewer provides options to change color of the bounding box and the viewer background but there is no user interface to do so.

## Solution

Add a section after Global Rendering Adjustments called "Customization" with a row for a color picker swatch for Background Color and another row for Bounding Box Color

## Acceptance Criteria

- Color pickers are functional and change color of background and bounding box lines, respectively.  Alpha is not important for either of the two colors so it would be good if the color picker were configured to not show alpha.
- Since bounding box can be turned on/off from top toolbar, in-panel color setting for bounding box should be disabled when bounding box button is off, which text that states, "bounding box turned off".

## Notes

We have a particular color picker library used in other parts of the app which is very configurable.  Look at how the channel colors are changed as a starting point.  It may be possible to use the identical color picker layout but without the alpha part of it.

## Design Specs
https://xd.adobe.com/view/904d6444-a800-48b6-8844-4f6f116d85ab-c51e/

### PR DESCRIPTION
**Pull request by frasercl**
Resolves #31:
- Adds a new "Customize" section with color pickers for background and bounding box color, which update dynamically
- Adds the props `backgroundColor` and `boundingBoxColor` to the `App` component
- Bounding box color picker is aware of whether the bounding box is visible, and will display a message if it is not

Modifies the `ColorPicker` component:
- Adds the prop `disableAlpha` to hide the alpha slider (which just passes through to the same prop on `SketchPicker`)
- Because the Customize section is so close to the bottom of its container, color pickers would usually appear off the bottom of the screen and be inaccessible. Therefore, `ColorPicker` is now aware of its position in screen space, and will open above the swatch if it does not have enough space to open below (see second screenshot). This change affects channel color pickers as well.
  - If this feature stays, `ColorPicker` should probably be able to receive a container to measure against, as it is still possible for color pickers to open off the bottom edge of their container if there is technically enough room on screen.

Misc:
- Removes the `reactcss` dependency (still required by `react-color`)
- Bumps `volume-viewer` version to get `view3d.setBackgroundColor` (see allen-cell-animated/volume-viewer#54)
- Changes the type of `channelSettings.color` in `UserSelectionState`, which I believe was mistyped

Screenshots from Cell Feature Explorer:
image
image

### CODE REVIEW

File: src/aics-image-viewer/components/App/index.tsx [hunk 1]
```diff
@@ -56,6 +56,7 @@ import {
 } from "../../shared/utils/sliderValuesToImageValues";
 
 import "./styles.css";
+import { colorArrayToFloatArray, rgbObjectToArray } from "../../shared/utils/colorObjectArrayConverting";
```
Comment (toloudis) [inline src/aics-image-viewer/components/App/index.tsx:59]: move this up to be next to the other utils import

File: src/aics-image-viewer/components/App/index.tsx [hunk 2]
```diff
@@ -438,6 +445,8 @@ export default class App extends React.Component<AppProps, AppState> {
     view3d.setCameraMode(enums.viewMode.VIEW_MODE_ENUM_TO_LABEL_MAP.get(userSelections.mode));
     view3d.setShowAxis(userSelections[SHOW_AXES]);
     view3d.setShowBoundingBox(aimg, userSelections.showBoundingBox);
+    view3d.setBoundingBoxColor(aimg, colorArrayToFloatArray(userSelections.boundingBoxColor));
+    view3d.setBackgroundColor(colorArrayToFloatArray(userSelections.backgroundColor));
```
Comment (toloudis) [inline src/aics-image-viewer/components/App/index.tsx:449]: since this (bg color) doesn't depend on aimg, it probably should not be called here.

File: src/aics-image-viewer/components/App/index.tsx [hunk 3]
```diff
@@ -719,6 +728,8 @@ export default class App extends React.Component<AppProps, AppState> {
     view3d.setCameraMode(enums.viewMode.VIEW_MODE_ENUM_TO_LABEL_MAP.get(userSelections.mode));
     view3d.setShowAxis(userSelections[SHOW_AXES]);
     view3d.setShowBoundingBox(aimg, userSelections.showBoundingBox);
+    view3d.setBackgroundColor(colorArrayToFloatArray(userSelections.backgroundColor));
+    view3d.setBoundingBoxColor(colorArrayToFloatArray(userSelections.boundingBoxColor));
```
Comment (toloudis) [inline src/aics-image-viewer/components/App/index.tsx:732]: setBoundingBoxColor needs to pass in aimg ?

File: src/aics-image-viewer/components/App/index.tsx [hunk 4]
```diff
@@ -719,6 +728,8 @@ export default class App extends React.Component<AppProps, AppState> {
     view3d.setCameraMode(enums.viewMode.VIEW_MODE_ENUM_TO_LABEL_MAP.get(userSelections.mode));
     view3d.setShowAxis(userSelections[SHOW_AXES]);
     view3d.setShowBoundingBox(aimg, userSelections.showBoundingBox);
+    view3d.setBackgroundColor(colorArrayToFloatArray(userSelections.backgroundColor));
```
Comment (toloudis) [inline src/aics-image-viewer/components/App/index.tsx:731]: same comment as above, bg color probably should be initialized prior to image load time

File: src/aics-image-viewer/components/ColorPicker.jsx [hunk 5]
```diff
@@ -58,45 +68,55 @@ class ColorPicker extends React.Component {
...
-          zIndex: '9999',
-        },
-        cover: {
-          position: 'fixed',
-          top: '0px',
-          right: '0px',
-          bottom: '0px',
-          left: '0px'
-        }
+    let styles = {
```
Comment (toloudis) [inline src/aics-image-viewer/components/ColorPicker.jsx:71]: should still be const?

File: src/aics-image-viewer/components/App/index.tsx (hunk 4 above)
Comment (toloudis) [reply src/aics-image-viewer/components/App/index.tsx:731]: suggestion: initialize bg color in CellViewerCanvasWrapper/index.jsx, in componentWillReceiveProps (pass bg color in as a prop to this component).   It may also need to be set in componentDidMount also...

File: src/aics-image-viewer/components/App/index.tsx (hunk 3 above)
Comment (frasercl) [reply src/aics-image-viewer/components/App/index.tsx:732]: looks like just about all the setters in volume-viewer with a `volume` argument (the ones that require `aimg`) leave that argument unused and instead check `this.image`, potentially to future-proof against the viewer eventually supporting multiple images at a time? if so, `setBoundingBoxColor` will need to receive a volume to know which of multiple bounding boxes to recolor.

File: src/aics-image-viewer/components/App/index.tsx [hunk 6]
```diff
@@ -103,6 +104,8 @@ const defaultProps: AppProps = {
     showBoundingBox: false,
     view: "3D", // "XY", "XZ", "YZ"
     mode: "default", // "pathtrace", "maxprojection"
+    backgroundColor: [0, 0, 0],
+    boundingBoxColor: [255, 255, 255],
```
Comment (meganrm) [inline src/aics-image-viewer/components/App/index.tsx:108]: These seem like they should be in constants so you can have a default you use other places

File: src/aics-image-viewer/components/App/index.tsx [hunk 7]
```diff
@@ -178,6 +181,8 @@ export default class App extends React.Component<AppProps, AppState> {
         [AUTO_ROTATE]: false,
         [SHOW_AXES]: props.viewerConfig.showAxes,
         showBoundingBox: props.viewerConfig.showBoundingBox,
+        boundingBoxColor: props.viewerConfig.boundingBoxColor || [255, 255, 255],
+        backgroundColor: props.viewerConfig.backgroundColor || [0, 0, 0],
```
Comment (meganrm) [inline src/aics-image-viewer/components/App/index.tsx:185]: Here's where you could use those default constants

File: src/aics-image-viewer/components/ControlPanel/index.jsx [hunk 8]
```diff
@@ -223,11 +224,11 @@ export default class ControlPanel extends React.Component {
...
-          title={[
-            this.renderAxesButton(),
-            this.renderBoundingBoxButton(),
-            this.renderResetCameraButton()
-          ]}
+          title={<>
+            {this.renderAxesButton()}
+            {this.renderBoundingBoxButton()}
+            {this.renderResetCameraButton()}
+          </>}
```
Comment (meganrm) [inline src/aics-image-viewer/components/ControlPanel/index.jsx:231]: what's the reason for this change?

File: src/aics-image-viewer/components/App/index.tsx (hunk 3 above)
Comment (toloudis) [reply src/aics-image-viewer/components/App/index.tsx:732]: yes, setBoundingBoxColor should get an aimg passed in for sure.

**Comment by lynwilhelm** [comment]
@frasercl  I would consider these couple things:
- Investigate if the bottom line in panel can be removed
- See if the 1px stroke around the color rectangles can be implemented globally without too much effort

Great work:)

File: src/aics-image-viewer/components/ControlPanel/index.jsx (hunk 8 above)
Comment (frasercl) [reply src/aics-image-viewer/components/ControlPanel/index.jsx:231]: React was complaining that this array of components did not have indexes; fragments don't have that requirement. I'm pretty sure this isn't the best practice (I believe fragments are meant to be used pretty much exclusively as the top-level container in JSX, to allow a component's `render` function to return multiple elements) but these buttons are soon to be moved out of the control panel anyways and into the new toolbar.

File: src/aics-image-viewer/components/App/index.tsx (hunk 4 above)
Comment (frasercl) [reply src/aics-image-viewer/components/App/index.tsx:731]: In this case `setShowAxis` should probably move as well. Looks like everything else that's here and doesn't accept `aimg` as an argument still internally depends on the image having been initialized (looking at `updateExposure`, `setCameraMode`).

Comment (toloudis) [reply src/aics-image-viewer/components/App/index.tsx:731]: basically agree, but my thinking was that the axes wont be indicative of much without an image there

Comment (frasercl) [reply src/aics-image-viewer/components/App/index.tsx:731]: makes sense, though in an editor configured to display axes by default, having the axes appear with no image at least indicates that the viewer has loaded correctly and any problem is with image loading.

File: src/aics-image-viewer/components/ColorPicker.jsx (hunk 5 above)
Comment (frasercl) [reply src/aics-image-viewer/components/ColorPicker.jsx]: was adding a property to the style based on whether the color picker should open above or below swatch, which I agree was weird. just changed to be more consistent with other single-file components, with an all-caps `const STYLES` at the very bottom.

Comment (toloudis) [reply src/aics-image-viewer/components/ColorPicker.jsx]: Adding a property shouldn't have changed constness.   Would be surprised if you got some tool-based indication that you had to change it.  Unless you were reassigning the `styles` variable itself.

Comment (frasercl) [reply src/aics-image-viewer/components/ColorPicker.jsx]: I was conditionally adding either one property or another after declaration. You're right that with the way const works in JS it wouldn't have mattered, but in personal projects I tend to change the keyword to indicate whether any property of an element changes, and I hadn't reconsidered that habit until just now. Regardless, style-wise I feel better about the changes I just made.

File: src/aics-image-viewer/components/ColorPicker.jsx [hunk 9]
```diff
@@ -58,48 +68,56 @@ class ColorPicker extends React.Component {
...
-          position: 'fixed',
-          top: '0px',
-          right: '0px',
-          bottom: '0px',
-          left: '0px'
-        }
-      }
-    });
+
+    let popoverDirectionStyle = {};
```
Comment (toloudis) [inline src/aics-image-viewer/components/ColorPicker.jsx:72]: const. 
could even be 
```
const popoverDirectionStyle = this.state.openAboveSwatch ? {bottom:"25px"} : {top:"1px"};
```
I suspect, because of the single quotes around the px values, that you need to run `npm run lint`, `npm run typeCheck`, and run prettier (formatter) to do the basic formatting checks.
We should probably see if there's a way to ensure autoformatting on commit or something.  I have "format on save" enabled for the project in vscode settings.

Comment (toloudis) [reply src/aics-image-viewer/components/ColorPicker.jsx:72]: Actually I see that the gh-actions did run the lint but I'm surprised it didn't warn about let vs const. I guess volume-viewer warns about this but website-3d-cell-viewer doesn't, in the eslint config ( "prefer-const": ["warn"] ).
I would advocate for adding this but in a separate maintenance pr because it might trigger a whole bunch of warnings.

File: src/aics-image-viewer/components/App/index.tsx [hunk 10]
```diff
@@ -103,6 +106,8 @@ const defaultProps: AppProps = {
     showBoundingBox: false,
     view: "3D", // "XY", "XZ", "YZ"
     mode: "default", // "pathtrace", "maxprojection"
+    backgroundColor: BACKGROUND_COLOR as [number, number, number],
+    boundingBoxColor: BOUNDING_BOX_COLOR as [number, number, number],
```
Comment (toloudis) [inline src/aics-image-viewer/components/App/index.tsx:110]: i suggest moving the [number,number,number] typing over to where the variable is defined rather than here.  you might have to rename the file to be a ts rather than a js file?

Comment (toloudis) [reply src/aics-image-viewer/components/App/index.tsx:110]: we can be gradually walking this codebase from js to ts overall

**Review by toloudis — APPROVED** [review]
Looking good.  I feel like the bg color should be passed to the CellViewerCanvasWrapper as a prop but it's not a deal breaker to do it this way at all.
* The typescript [number,number,number] is the only thing I would still change to make this code cleaner.

File: src/aics-image-viewer/components/ColorPicker.jsx (hunk 9 above)
Comment (frasercl) [reply src/aics-image-viewer/components/ColorPicker.jsx:72]: Yep, your way is better on this. Don't know how I missed it in the first place. And eslint throws up ~50 warnings when told to care about const, so definitely an issue for another time.
